### PR TITLE
format-table: refac texture formats

### PR DESCRIPTION
**Pull request by macabeus**
Closes https://github.com/Ryujinx/Ryujinx/issues/1889

It's a suggestion by @gdkchan ([message link](https://discord.com/channels/410208534861447168/410432399500115968/797557576727265330))

I wrote the following JS script to generate the code that I used on the PR. It's a very simple code, so I just wrote a very quick and dirty code.
  ```javascript
const map = {
    '0011101': 'R8',
    '0011011': 'R16',
    '0001111': 'R32',
    '0011000': 'G8R8',
    '0001100': 'R16G16',
    '0000100': 'R32G32',
    '0000010': 'R32G32B32',
    '0001000': 'A8B8G8R8',
    '0000011': 'R16G16B16A16',
    '0000001': 'R32G32B32A32',
    '0111010': 'X8Z24X20V4S8COV4R4V',
    '0101111': 'Zf32',
    '0001110': 'G24R8',
    '0101001': 'Z24S8',
    '0000101': 'R32B24G8',
    '0110000': 'Zf32X24S8',
    '0010010': 'A4B4G4R4',
    '0010100': 'A1B5G5R5',
    '0010101': 'B5G6R5',
    '0001001': 'A2B10G10R10',
    '0100001': 'Bf10GF11RF11',
    '0100000': 'E5B9G9R9SHaredexp',
    '0100100': 'Dxt1',
    '0100101': 'Dxt23',
    '0100110': 'Dxt45',
    '0010011': 'A5B5G5R1',
    '0100111': 'Dxn1',
    '0101000': 'Dxn2',
    '0010111': 'Bc7U',
    '0001011': 'Etc2Rgba',
    '0010000': 'Rtdoublebind',
    '0010001': 'Rttypesmismatch',
    '1000000': 'Astc2D4x4',
    '1010000': 'Astc2D5x4',
    '1000001': 'Astc2D5x5',
    '1010001': 'Astc2D6x5',
    '1000010': 'Astc2D6x6',
    '1010101': 'Astc2D8x5',
    '1010010': 'Astc2D8x6',
    '1000100': 'Astc2D8x8',
    '1010110': 'Astc2D10x5',
    '1010111': 'Astc2D10x6',
    '1010011': 'Astc2D10x8',
    '1000101': 'Astc2D10x10',
    '1010100': 'Astc2D12x10',
    '1000110': 'Astc2D12x12',
    '0101010': 'X8Z24',
    '0101011': 'S8Z24',
    '0100010': 'G8B8G8R8',
    '0100011': 'B8G8R8G8',

    '001': 'Snorm',
    '010': 'Unorm',
    '011': 'Sint',
    '100': 'Uint',
    '101': 'SNormForceFP16',
    '111': 'Float',
}

function checkMissing(value, type) {
    if (map[value] === undefined) {
        console.log(`Missing ${type} "${value}" (0x${parseInt(value, 2).toString(16)})`)
        return true
    }
}

function main(hex) {
    [, srgb, alpha, blue, green, red, size] = parseInt(hex, 16)
        .toString(2)
        .padStart(20, '0')
        .match(/(.)(...)(...)(...)(...)(.{7})$/)

    if (checkMissing(size, 'size')) { return }
    if (checkMissing(red, 'red')) { return }
    if (checkMissing(green, 'green')) { return }
    if (checkMissing(blue, 'blue')) { return }
    if (checkMissing(alpha, 'alpha')) { return }

    const srgbNameAppend = (
        srgb === '0'
        ? ''
        : 'Srgb'
    )

    const srgbAppend = (
        srgb === '0'
        ? ''
        : ' | Srgb'
    )

    if (red === green && green === blue) {
        return `${map[size]}${map[red]}${srgbNameAppend} = ${map[size]} | R${map[red]} | G${map[green]} | B${map[blue]} | A${map[alpha]}${srgbAppend}, // ${hex}`
    }

    return `= ${map[size]} | R${map[red]} | G${map[green]} | B${map[blue]} | A${map[alpha]}${srgbAppend}, // ${hex}`
}
```

You can copy and paste it on your browser's console and then use with:

```js
copy(
[
    '0x7ff90',
    '0x7ff91',
].map(main).join('\n')
)
```

### CODE REVIEW

File: Ryujinx.Graphics.Gpu/Image/FormatTable.cs [hunk 1]
```diff
@@ -8,107 +8,292 @@ namespace Ryujinx.Graphics.Gpu.Image
...
+            // DXN2Snorm_Second              = DXN2                | RSnorm | GSnorm | BSnorm | ASNormForceFP16, // 0xa4951
+            BF10GF11RF11Snorm             = BF10GF11RF11        | RSnorm | GSnorm | BSnorm | ASNormForceFP16, // 0xa4942
+            X8Z24Snorm_First              = X8Z24               | RSnorm | GSnorm | BSnorm | ASNormForceFP16, // 0xa4955
+            Z24S8Snorm_First              = Z24S8               | RSnorm | GSnorm | BSnorm | ASNormForceFP16, // 0xa4952
+            G8B8G8R8Snorm_First           = G8B8G8R8            | RSnorm | GSnorm | BSnorm | ASNormForceFP16, // 0xa4944
+            S8Z24Snorm_First              = S8Z24               | RSnorm | GSnorm | BSnorm | ASNormForceFP16, // 0xa4956
+            // S8Z24Snorm_Second             = S8Z24               | RSnorm | GSnorm | BSnorm | ASNormForceFP16, // 0xa4957
+            // Z24S8Snorm_Second             = Z24S8               | RSnorm | GSnorm | BSnorm | ASNormForceFP16, // 0xa4953
+            // G8B8G8R8Snorm_Second          = G8B8G8R8            | RSnorm | GSnorm | BSnorm | ASNormForceFP16, // 0xa4945
```
Comment (macabeus) [inline Ryujinx.Graphics.Gpu/Image/FormatTable.cs:190]: There are cases, like this one, with the hex value is different, but the calc is the same!
```
G8B8G8R8Snorm_First           = G8B8G8R8            | RSnorm | GSnorm | BSnorm | ASNormForceFP16, // 0xa4944
G8B8G8R8Snorm_Second          = G8B8G8R8            | RSnorm | GSnorm | BSnorm | ASNormForceFP16, // 0xa4945
```

@gdkchan How to solve that? Just adding a `+ 1` at the end?

Comment (gdkchan) [reply Ryujinx.Graphics.Gpu/Image/FormatTable.cs:190]: This looks incorrect, they are not this format on the XML, both are ASTC formats with different block sizes.

Comment (macabeus) [reply Ryujinx.Graphics.Gpu/Image/FormatTable.cs:190]: @gdkchan Okay. I found the problem. On my script, I'm getting the firsts digits, not the lasts.

**Review by gdkchan — COMMENTED** [review]
Looks good. For now I would suggest the following improvements:
- Naming on the `GpuFormat` enum should match the `Format` naming as close as possible. Mostly need to change casing here, like `Srgb` rather than `SRGB`, `Astc2D4x4Unorm` rather than `ASTC2D4X4Unorm` etc.
- I think the commend with the hex value of the enum entry at the end of the line on the `GpuFormat` is not necessary, it will make adding new formats there more complicated as one will need to calculate this value.
- We don't align the commas on the table in other files. The alignment is fine, its just that we usually do this:
```cs
ShortName    = 1,   2,   3,
VeryLongName = 123, 456, 789,
```
rather than:
```cs
ShortName    = 1   , 2   , 3   ,
VeryLongName = 123 , 456 , 789 ,
```
- Rather than casting to uint on the dictionary initializer, I think its better to change the key type to `GpuFormat`. The method that retrieves the format from the dictionary can cast from `uint` to `GpuFormat`.

**Comment by macabeus** [comment]
@gdkchan Done.

File: Ryujinx.Graphics.Gpu/Image/FormatTable.cs [hunk 2]
```diff
@@ -8,107 +8,294 @@ namespace Ryujinx.Graphics.Gpu.Image
...
+            G24R8               = 0x0E,
+            Z24S8               = 0x29,
+            R32B24G8            = 0x05,
+            Zf32X24S8           = 0x30,
+            A4B4G4R4            = 0x12,
+            A1B5G5R5            = 0x14,
+            B5G6R5              = 0x15,
+            A2B10G10R10         = 0x09,
+            Bf10GF11RF11        = 0x21,
+            E5B9G9R9SHaredexp   = 0x20,
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Image/FormatTable.cs:34]:
```suggestion
            E5B9G9R9SharedExp   = 0x20,
```

File: Ryujinx.Graphics.Gpu/Image/FormatTable.cs [hunk 3]
```diff
@@ -8,107 +8,294 @@ namespace Ryujinx.Graphics.Gpu.Image
...
+            Dxt1                = 0x24,
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Image/FormatTable.cs:35]:
```suggestion
            Bc1                 = 0x24,
```

File: Ryujinx.Graphics.Gpu/Image/FormatTable.cs [hunk 4]
```diff
@@ -8,107 +8,294 @@ namespace Ryujinx.Graphics.Gpu.Image
...
+            Dxt23               = 0x25,
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Image/FormatTable.cs:36]:
```suggestion
            Bc2                 = 0x25,
```

File: Ryujinx.Graphics.Gpu/Image/FormatTable.cs [hunk 5]
```diff
@@ -8,107 +8,294 @@ namespace Ryujinx.Graphics.Gpu.Image
...
+            Dxt45               = 0x26,
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Image/FormatTable.cs:37]:
```suggestion
            Bc3                 = 0x26,
```

File: Ryujinx.Graphics.Gpu/Image/FormatTable.cs [hunk 6]
```diff
@@ -8,107 +8,294 @@ namespace Ryujinx.Graphics.Gpu.Image
...
+            A5B5G5R1            = 0x13,
+            Dxn1                = 0x27,
+            Dxn2                = 0x28,
+            Bc7U                = 0x17,
+            Etc2Rgba            = 0x0B,
+            Rtdoublebind        = 0x10,
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Image/FormatTable.cs:43]: Use the Format enum name here (`Bc6HSfloat`).

File: Ryujinx.Graphics.Gpu/Image/FormatTable.cs [hunk 7]
```diff
@@ -8,107 +8,294 @@ namespace Ryujinx.Graphics.Gpu.Image
...
+            Rttypesmismatch     = 0x11,
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Image/FormatTable.cs:44]: Use Format enum name here (`Bc6HUfloat`).

File: Ryujinx.Graphics.Gpu/Image/FormatTable.cs [hunk 8]
```diff
@@ -8,107 +8,294 @@ namespace Ryujinx.Graphics.Gpu.Image
...
+            A1B5G5R5            = 0x14,
+            B5G6R5              = 0x15,
+            A2B10G10R10         = 0x09,
+            Bf10GF11RF11        = 0x21,
+            E5B9G9R9SHaredexp   = 0x20,
+            Dxt1                = 0x24,
+            Dxt23               = 0x25,
+            Dxt45               = 0x26,
+            A5B5G5R1            = 0x13,
+            Dxn1                = 0x27,
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Image/FormatTable.cs:39]: Use format enum name (`Bc4`).

File: Ryujinx.Graphics.Gpu/Image/FormatTable.cs [hunk 9]
```diff
@@ -8,107 +8,294 @@ namespace Ryujinx.Graphics.Gpu.Image
...
+            Dxn2                = 0x28,
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Image/FormatTable.cs:40]: Use format enum name (`Bc5`).

**Review by gdkchan — COMMENTED** [review]
The changes should also be applied to the entries below with the `|`s.

**Comment by macabeus** [comment]
@gdkchan Done.

File: Ryujinx.Graphics.Gpu/Image/FormatTable.cs [hunk 10]
```diff
@@ -8,107 +8,294 @@ namespace Ryujinx.Graphics.Gpu.Image
...
+            Astc2D10x8UnormSrgb      = Astc2D10x8          | RUnorm | GUnorm | BUnorm | AUnorm | Srgb,
+            Astc2D10x10UnormSrgb     = Astc2D10x10         | RUnorm | GUnorm | BUnorm | AUnorm | Srgb,
+            Astc2D12x10UnormSrgb     = Astc2D12x10         | RUnorm | GUnorm | BUnorm | AUnorm | Srgb,
+            Astc2D12x12UnormSrgb     = Astc2D12x12         | RUnorm | GUnorm | BUnorm | AUnorm | Srgb,
+            A5B5G5R1Unorm            = A5B5G5R1            | RUnorm | GUnorm | BUnorm | AUnorm,
+        }
+
+        private static Dictionary<GpuFormat, FormatInfo> _textureFormats = new Dictionary<GpuFormat, FormatInfo>()
+        {
+            { GpuFormat.R8Unorm,                   new FormatInfo(Format.R8Unorm,              1,   1,   1, 1) },
```
Comment (AcK77) [inline Ryujinx.Graphics.Gpu/Image/FormatTable.cs:200]: There is some extra spaces in the alignment in childs which could be removed!

File: Ryujinx.Graphics.Gpu/Image/FormatTable.cs [hunk 11]
```diff
@@ -205,7 +392,7 @@ public static bool TryGetTextureFormat(uint encoded, bool isSrgb, out FormatInfo
         {
             encoded |= (isSrgb ? 1u << 19 : 0u);
 
-            return _textureFormats.TryGetValue(encoded, out format);
+            return _textureFormats.TryGetValue((GpuFormat) encoded, out format);
```
Comment (gdkchan) [inline Ryujinx.Graphics.Gpu/Image/FormatTable.cs:395]:
```suggestion
            return _textureFormats.TryGetValue((GpuFormat)encoded, out format);
```

**Review by gdkchan — COMMENTED** [review]
Looks good other than what was already commented.

**Comment by Shahil-Ayato** [comment]
Needs a rebase.

**Comment by AcK77** [comment]
@macabeus are you still interested to working on this?

**Comment by macabeus** [comment]
@AcK77 I'm focusing only on the website now, so I'm not working on this PR anymore

**Comment by AcK77** [comment]
Closed due to inactivity

**Pull request closed without merging**